### DISCUSSION TITLE
how are constant monomial aux variables computed from material properties?

### DISCUSSION THREAD

**Discussion by jessecarterMOOSE** [idaholab/moose#23892] [Q&A General] [answered]
How are these calculated? Is it just a simple algebraic mean (sum(values)/N) over the qp's on the element? Does the element volume come into play? Or is there some kind of integral involving a shape function?

Specifically I'm using the Tensor Mechanics Master Action to generate constant monomial auxvariables from stresses, but those are just  material properties so the same logic should apply I'm guessing?

**Comment by hugary1995** [comment, marked as answer]
It's a volumetric averaging, i.e. you integrate the material property over the element volume then divide by the element volume.
```
    _n_local_dofs = _var.numberOfDofs();
    if (_n_local_dofs == 1) /* p0 */
    {
      ComputeValueType value = 0;
      for (_qp = 0; _qp < _qrule->n_points(); _qp++)
        value += _JxW[_qp] * _coord[_qp] * computeValue();
      value /= (_bnd ? _current_side_volume : _current_elem_volume);
      if (_var.isFV())
        setDofValueHelper(value);
      else
        // update the variable data referenced by other kernels.
        // Note that this will update the values at the quadrature points too
        // (because this is an Elemental variable)
        _var.setNodalValue(value);
    }
```

**Reply by jessecarterMOOSE** [reply]
What kind of quadrature rule is involved for this?

**Reply by hugary1995** [reply]
That's actually a good question. Ideally, one should be able to pick a quadrature rule. But MOOSE has taken the monopoly decision here -- all objects (kernels, auxkernels, materials, userobjects, etc.) use the same quadrature rule.

The default quadrature rule should be Gauss quadrature, the order is determined by the maximum variable order.

**Reply by GiudGiud** [reply]
The quadrature defined in the executioner. By default GAUSS with an order sufficient to integrate exactly the shape function * test function (polynomials) products

**Reply by jessecarterMOOSE** [reply]
so would it be the same quadrature used for the nonlinear system variables?
